File: frameworks/rs/tests/java_api/SSHealingBrush/src/rs/example/android/com/healingbrush/Healing.java
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package rs.example.android.com.healingbrush;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.Rect;
import android.renderscript.Allocation;
import android.renderscript.Element;
import android.renderscript.RenderScript;
import android.renderscript.Script;
import android.renderscript.Type;
import android.util.Log;

import com.example.android.rs.sample.ScriptC_healing;

public class Healing {
    private static final String TAG = "Healing";
    Rect mRoiBounds; // bounding box of the ROI
    float[] mPointsXY; // polygon point in original image coordnates
    int mCutOffsetX; // image coords of the cut  (mPointsXY - mPasteOffX + mCutOffsetX)
    int mCutOffsetY; // image coords of the cut (mPointsXY - mPasteOffY + mCutOffsetY)
    Bitmap mUndoBitmap;

    public Healing(Rect roiBounds, float[] pointsXY, int cutOffsetX, int cutOffsetY) {
        mRoiBounds = roiBounds;
        mPointsXY = pointsXY;
        mCutOffsetX = cutOffsetX;
        mCutOffsetY = cutOffsetY;
    }

    private static Bitmap buildMask(Rect rec, float[] xy) {
        Bitmap bitmap = Bitmap.createBitmap(rec.width(), rec.height(), Bitmap.Config.ALPHA_8);

        Canvas c = new Canvas(bitmap);
        Paint paint = new Paint();
        paint.setStyle(Paint.Style.FILL);
        paint.setColor(Color.BLACK);
        Path path = new Path();
        for (int i = 0; i < xy.length; i += 2) {
            if (i == 0) {
                path.moveTo(xy[i] - rec.left, xy[i + 1] - rec.top);
            } else {
                path.lineTo(xy[i] - rec.left, xy[i + 1] - rec.top);
            }
        }
        path.close();
        c.drawPath(path, paint);
        return bitmap;
    }

    /**
     * This function only assumes mPointsXY, mPasteOffX, mPasteOffY
     *
     * @param healing
     * @param rs
     * @param image
     */
    public void heal(ScriptC_healing healing, RenderScript rs, Bitmap image, Bitmap output) {
        long time = System.nanoTime();

        Type.Builder floatImage = new Type.Builder(rs, Element.F32_3(rs));
        floatImage.setX(mRoiBounds.width());
        floatImage.setY(mRoiBounds.height());


        Bitmap mask_bitmap = buildMask(mRoiBounds, mPointsXY);
        Bitmap dest_bitmap = createMutableBitmap(image, mRoiBounds.left, mRoiBounds.top,
                mRoiBounds.width(), mRoiBounds.height());
        Allocation dest_alloc = Allocation.createFromBitmap(rs, dest_bitmap);
        Bitmap src_bitmap = createMutableBitmap(image, mCutOffsetX, mCutOffsetY,
                mRoiBounds.width(), mRoiBounds.height());
        Allocation src_alloc = Allocation.createFromBitmap(rs, src_bitmap);
        Allocation mask_alloc = Allocation.createFromBitmap(rs, mask_bitmap);

        healing.invoke_heal(mask_alloc, src_alloc, dest_alloc);

        dest_alloc.copyTo(dest_bitmap);

        dest_bitmap.setHasAlpha(true);

        // build the undo
        mUndoBitmap = Bitmap.createBitmap(mRoiBounds.width(), mRoiBounds.height(),
                Bitmap.Config.ARGB_8888);
        Canvas undoCanvas = new Canvas(mUndoBitmap);
        Rect undoRect = new Rect(0, 0, mRoiBounds.width(), mRoiBounds.height());
        undoCanvas.drawBitmap(output, mRoiBounds, undoRect, null);

        Canvas c = new Canvas(output);
        c.drawBitmap(image, 0, 0, null);
        c.drawBitmap(dest_bitmap, mRoiBounds.left, mRoiBounds.top, null);
        Log.v(TAG, " time ss to smart paste = " + (System.nanoTime() - time) / 1E6f + "ms");
        heal_orig(healing, rs, image, output);

    }


    /**
     * This function only assumes mPointsXY, mPasteOffX, mPasteOffY
     *
     * @param healing
     * @param rs
     * @param image
     */
    public void heal_orig(ScriptC_healing healing, RenderScript rs, Bitmap image, Bitmap output) {
        long time = System.nanoTime();
        Type.Builder floatImage = new Type.Builder(rs, Element.F32_3(rs));
        floatImage.setX(mRoiBounds.width());
        floatImage.setY(mRoiBounds.height());

        Bitmap maskBitmap = buildMask(mRoiBounds, mPointsXY);

        Allocation dest1 = Allocation.createTyped(rs, floatImage.create());
        Allocation dest2 = Allocation.createTyped(rs, floatImage.create());
        healing.set_dest1(dest1);
        healing.set_dest2(dest2);

        Bitmap destBitmap = createMutableBitmap(image, mRoiBounds.left, mRoiBounds.top,
                mRoiBounds.width(), mRoiBounds.height());
        Allocation dest_uc4 = Allocation.createFromBitmap(rs, destBitmap);
        healing.forEach_convert_to_f(dest_uc4, dest1);

        Bitmap src = createMutableBitmap(image, mCutOffsetX, mCutOffsetY,
                mRoiBounds.width(), mRoiBounds.height());
        Allocation src_f3 = Allocation.createTyped(rs, floatImage.create());
        Allocation src_uc4 = Allocation.createFromBitmap(rs, src);
        healing.forEach_convert_to_f(src_uc4, src_f3);
        healing.set_src(src_f3);

        Allocation mask = Allocation.createFromBitmap(rs, maskBitmap);
        healing.set_mask(mask);

        Allocation laplace_f3 = Allocation.createTyped(rs, floatImage.create());
        healing.set_laplace(laplace_f3);

        Script.LaunchOptions options = new Script.LaunchOptions();
        options.setX(1, mRoiBounds.width() - 1);
        options.setY(1, mRoiBounds.height() - 1);
        healing.forEach_laplacian(laplace_f3, options);
        healing.forEach_copyMasked(mask, dest1);

        int steps = (int) Math.hypot(mRoiBounds.width(), mRoiBounds.height()); // match RS Single source
        Log.v(TAG, "Healing_orig  :steps = " + steps);
        for (int i = 0; i < steps; i++) {
            healing.forEach_solve1(mask, dest2);
            healing.forEach_solve2(mask, dest1);
        }

        healing.forEach_convert_to_uc(dest1, dest_uc4);
        rs.finish();

        healing.forEach_alphaMask(dest_uc4, dest_uc4);
        rs.finish();

        dest_uc4.copyTo(destBitmap);
        rs.finish();
        destBitmap.setHasAlpha(true);
        rs.finish();
        // build the undo
        mUndoBitmap = Bitmap.createBitmap(mRoiBounds.width(), mRoiBounds.height(),
                Bitmap.Config.ARGB_8888);
        Canvas undoCanvas = new Canvas(mUndoBitmap);
        Rect undoRect = new Rect(0, 0, mRoiBounds.width(), mRoiBounds.height());
        undoCanvas.drawBitmap(output, mRoiBounds, undoRect, null);

        Canvas c = new Canvas(output);
        c.drawBitmap(image, 0, 0, null);
        c.drawBitmap(destBitmap, mRoiBounds.left, mRoiBounds.top, null);
        Log.v(TAG, " time to smart paste = " + (System.nanoTime() - time) / 1E6f + "ms");
    }

    Bitmap createMutableBitmap(Bitmap image, int x, int y, int width, int height) {
        Bitmap ret = Bitmap.createBitmap(image, x, y, width, height);
        return ret.copy(Bitmap.Config.ARGB_8888, true);
    }

    private static int calcMaskArea(Allocation mask) {
        int w = mask.getType().getX();
        int h = mask.getType().getY();
        byte[] data = new byte[w * h];
        mask.copyTo(data);
        int count = 0;
        int val = data[0];
        for (int i = 0; i < data.length; i++) {
            if (data[i] != val) {
                count++;
            }
        }
        return count;
    }

    Bitmap getmUndoBitmap() {
        return mUndoBitmap;
    }
}

File: frameworks/rs/tests/java_api/SSHealingBrush/src/rs/example/android/com/healingbrush/healing.rs
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#pragma version(1)
#pragma rs java_package_name(com.example.android.rs.sample)
#pragma rs_fp_relaxed

typedef rs_allocation AllocationF32_3;

AllocationF32_3 src;

float3 __attribute__((kernel)) laplacian(uint32_t x, uint32_t y) {
  float3 out = 4 * rsGetElementAt_float3(src, x, y);
  out -= rsGetElementAt_float3(src, x - 1, y);
  out -= rsGetElementAt_float3(src, x + 1, y);
  out -= rsGetElementAt_float3(src, x, y - 1);
  out -= rsGetElementAt_float3(src, x, y + 1);
  return out;
}

rs_allocation mask;       // uchar
AllocationF32_3 laplace;  // float3
AllocationF32_3 dest1;    // float3
AllocationF32_3 dest2;    // float3

float3 __attribute__((kernel)) convert_to_f(uchar4 in) {
  return convert_float3(in.xyz);
}
float3 __attribute__((kernel)) copy(float3 in) { return in; }

float3 __attribute__((kernel)) copyMasked(uchar in, uint32_t x, uint32_t y) {
  return rsGetElementAt_float3((in > 0) ? src : dest1, x, y);
}

uchar4 __attribute__((kernel)) convert_to_uc(float3 in) {
  in = clamp(in, 0.0f, 255.0f);
  return convert_uchar4((float4){in.x, in.y, in.z, 0xFF});
}

uchar4 __attribute__((kernel)) alphaMask(uchar4 in, uint32_t x, uint32_t y) {
  if (rsGetElementAt_uchar(mask, x, y) == 0) {
    return (uchar4){0, 0, 0, 0};
  }

  return in;
}

float3 __attribute__((kernel)) solve1(uchar in, uint32_t x, uint32_t y) {
  if (in > 0) {
    float3 k = rsGetElementAt_float3(dest1, x - 1, y);
    k += rsGetElementAt_float3(dest1, x + 1, y);
    k += rsGetElementAt_float3(dest1, x, y - 1);
    k += rsGetElementAt_float3(dest1, x, y + 1);
    k += rsGetElementAt_float3(laplace, x, y);
    k /= 4;
    return k;
  }
  return rsGetElementAt_float3(dest1, x, y);
}

float3 __attribute__((kernel)) solve2(uchar in, uint32_t x, uint32_t y) {
  if (in > 0) {
    float3 k = rsGetElementAt_float3(dest2, x - 1, y);
    k += rsGetElementAt_float3(dest2, x + 1, y);
    k += rsGetElementAt_float3(dest2, x, y - 1);
    k += rsGetElementAt_float3(dest2, x, y + 1);
    k += rsGetElementAt_float3(laplace, x, y);
    k /= 4;
    return k;
  }
  return rsGetElementAt_float3(dest2, x, y);
}

rs_allocation image;
rs_allocation border;         // float3
rs_allocation border_coords;  // int2
int borderLength;

float3 __attribute__((kernel)) extractBorder(int2 in) {
  return convert_float3(rsGetElementAt_uchar4(image, in.x, in.y).xyz);
}

float __attribute__((kernel)) bordercorrelation(uint32_t x, uint32_t y) {
  float sum = 0;
  for (int i = 0; i < borderLength; i++) {
    int2 coord = rsGetElementAt_int2(border_coords, i);
    float3 orig = convert_float3(
        rsGetElementAt_uchar4(image, coord.x + x, coord.y + y).xyz);
    float3 candidate = rsGetElementAt_float3(border, i).xyz;
    sum += distance(orig, candidate);
  }
  return sum;
}

static inline rs_allocation toFloat3(rs_allocation in) {
  int width = rsAllocationGetDimX(in);
  int height = rsAllocationGetDimY(in);

  rs_allocation tmp = rsCreateAllocation_float3(width, height);
  rsForEach(convert_to_f, in, tmp);
  return tmp;
}

static rs_allocation clone(rs_allocation in) {
  int width = rsAllocationGetDimX(in);
  int height = rsAllocationGetDimY(in);
  rs_allocation tmp = rsCreateAllocation_float3(width, height);
  rsForEach(copy, in, tmp);
  return tmp;
}

void heal(rs_allocation mask_image, rs_allocation src_image,
          rs_allocation dest_image) {
  int width = rsAllocationGetDimX(src_image);
  int height = rsAllocationGetDimY(src_image);
  src = toFloat3(src_image);
  mask = mask_image;
  laplace = rsCreateAllocation_float3(width, height);
  dest1 = toFloat3(dest_image);
  dest2 = clone(dest1);

  int steps = (int)hypot((float)width, (float)height);
  rsDebug("find_region.rs:steps = ", steps);

  rs_script_call_t opts = {0};
  opts.xStart = 1;
  opts.xEnd = width - 1;
  opts.yStart = 1;
  opts.yEnd = height - 1;
  rsForEachWithOptions(laplacian, &opts, laplace);
  rsForEach(copyMasked, mask, dest1);
  for (int i = 0; i < steps; i++) {
    rsForEach(solve1, mask, dest2);
    rsForEach(solve2, mask, dest1);
  }

  rsForEach(convert_to_uc, dest1, dest_image);
  rsForEach(alphaMask, dest_image, dest_image);
}
